doc/ubertooth-rx.1
.TH UBERTOOTH\-RX 7 "March 2017" "Project Ubertooth" "User Commands"
.SH NAME
.PP
.BR ubertooth-rx (1) 
\- Classic Bluetooth discovery, sniffing, and decoding
.SH SYNOPSIS
.PP
.RS
.nf
ubertooth\-rx [ \-l <lap> [ \-u <uap ] ]
ubertooth\-rx \-z
.fi
.RE
.SH DESCRIPTION
.PP
.BR ubertooth-rx (1) 
is the primary interface into Classic Bluetooth (BR)
functionality provided by Ubertooth. It has two main modes of operation:
piconet following and survey mode. In either mode, 
.BR ubertooth-rx (1) 
is
able to discover undiscoverable devices. See [DISCOVERING UNDISCOVERABLE
DEVICES][].
.PP
In piconet following mode, the tool will follow the first piconet it
fully identifies. In survey mode the device will attempt to identify all
piconets in a given area and display them after either a timeout or
manual interruption.
.PP
Piconet following is the main mode entered when no arguments are passed
to the command or a LAP and optionally a UAP are provided. If no
arguments are passed, the tool will attempt to calculate the UAP for any
observed LAPs. If a LAP is passed, the UAP will be calculated for that
specific LAP. Once a LAP and UAP have been recovered, the tool will
attempt to recover the clock value, and if that succeeds it will follow
that piconet.
.PP
Survey mode, entered using \fB\fC\-z\fR, will record all LAPs and attempt to
calculate the UAPs for any observed LAPs. This mode can be combined with
a timeout using \fB\fC\-t\fR, and it can be interrupted at any time using
\fB\fCctrl\-C\fR\&.
.SH EXAMPLES
.PP
Follow the first piconet whose LAP, UAP, and clock are recovered from
the air:
.PP
.RS
.nf
ubertooth\-rx
.fi
.RE
.PP
For a given LAP, calculate the UAP and recover the clock, then follow:
.PP
.RS
.nf
ubertooth\-rx \-l 112233
.fi
.RE
.PP
For a given LAP and UAP, recover the clock then follow:
.PP
.RS
.nf
ubertooth\-rx \-l 112233 \-u ab
.fi
.RE
.PP
Enter survey mode for 20 seconds, and print out the BD ADDRs of all
observed piconets:
.PP
.RS
.nf
ubertooth\-rx \-z \-t 20
.fi
.RE
.SH OPTIONS
.PP
Major modes:
.RS
.IP \(bu 2
\fB\fC\-l <lap>\fR :
Limit UAP recovery, clock recovery, and piconet following to a given
LAP. Format is 3 bytes / 6 hex characters.
.IP \(bu 2
\fB\fC\-u <uap>\fR :
Limit clock recovery and piconet following to a given UAP. Must be
used in conjunction with \fB\fC\-l\fR\&. Format is 1 byte / 2 hex characters.
.IP \(bu 2
\fB\fC\-z\fR :
Survey mode: recover all LAP and UAP pairs and display them. Will run
indefinitely until interrupted with \fB\fCctrl\-C\fR unless paired with \fB\fC\-t\fR\&.
.RE
.PP
Options:
.RS
.IP \(bu 2
\fB\fC\-i <input>\fR :
Input file. If not specified will perform live capture using
Ubertooth.
.IP \(bu 2
\fB\fC\-c <0\-79>\fR :
Fixed channel for all major modes. If not specified will sweep
through all channels.
.IP \(bu 2
\fB\fC\-s\fR :
Sweep through channels in all major modes. See also: \fB\fC\-c\fR\&.
.IP \(bu 2
\fB\fC\-e <0\-4>\fR :
Maximum access code bit errors. [Default: 2]
.IP \(bu 2
\fB\fC\-t <seconds>\fR :
Timeout in seconds. If not specified will run indefinitely. Suggested
values for \fB\fC\-z\fR: 20\-60 seconds.
.RE
.PP
Output options:
.RS
.IP \(bu 2
\fB\fC\-r <file.pcapng>\fR :
Capture packets to PcapNG
.IP \(bu 2
\fB\fC\-q <file.pcap>\fR :
Capture packets to PCAP
.IP \(bu 2
\fB\fC\-d <file.bin>\fR :
Capture packets to binary file suitable for use with \fB\fC\-i\fR\&.
.RE
.PP
Miscellaneous:
.RS
.IP \(bu 2
\fB\fC\-V\fR :
Version information
.IP \(bu 2
\fB\fC\-U<0\-7>\fR :
Which Ubertooth device to use
.RE
.SH DISCOVERING UNDISCOVERABLE DEVICES
.PP
Classic Bluetooth piocnets are defined by the Lower Address Part (LAP)
and Upper Address Part (UAP) of the master device. These are elements
of the master device's Bluetooth Address (BD ADDR).
.PP
Consider the following BD ADDR:
.PP
.RS
.nf
22:44:66:88:AA:BB
.fi
.RE
.PP
The lower address part (LAP) is the lower 24 bits, so \fB\fC88:AA:BB\fR\&. In the
context of this tool, the value is written \fB\fC88AABB\fR\&. The upper address
part is the next 8 bits, so \fB\fC66\fR\&. The \fB\fC22:44\fR is called the
Non\-significant Address Part (NAP) and as you might imagine it is not
significant.
.PP
In piconet following mode, the tool will recover LAP values from the air
and attempt to calculate the UAP from those. It will go on to follow the
piconet if it can recover the clock value. In survey mode, the tool will
simply recover LAP and UAP values.
.PP
To convert LAP + UAP pairs back into Bluetooth addresses, do the reverse
of the above. For example, if the tool recovers a LAP of 36A2B4 and a
UAP of 98, the associated Bluetooth address is \fB\fC??:??:98:36:A2:B4\fR\&. Any
value can be substituted into the ?? slots and most Bluetooth tools will
still work. For example, \fB\fChcitool name 00:00:98:36:A2:B4\fR will establish
a connection to the device and return its name.
.PP
This attack works against discoverable and undiscoverable devices alike.
.SH SEE ALSO
.PP
.BR ubertooth-scan (1): 
active device scanning and inquiry using Ubertooth
and BlueZ
.PP
.BR ubertooth (7): 
overview of Project Ubertooth
.PP
D. Spill and A. Bittau. "BlueSniff: Eve Meets Alice and Bluetooth."
USENIX WOOT 2007.
.SH AUTHOR
.PP
This manual page was written by Mike Ryan.
.SH COPYRIGHT
.PP
.BR ubertooth-rx (1) 
is Copyright (c) 2010\-2017 Michael Ossmann, Dominic
Spill, and others. This tool is released under the GPLv2. Refer to
\fB\fCCOPYING\fR for further details.


changelog
ubertooth (2017-03-R2) unstable; urgency=low
  * USB API versioning - We have added USB API versioning to improve the
    experience of failures due to mismatched firmware and host code. You may
    find that Ubertooth tools complain that your firmware is out of date.
    Updating firmware is simple and reliable, please follow the instructions
    on the wiki. A release archive with precompiled firmware is available.
  * Basic Rate PCAP output - the pcap file produced by the Ubertooth tools
    should all be read by the latest Wireshark versions without the need to
    install plugins.
  * Kismet plugins - Kismet is currently under heavy development and it has
    not been possible to continue tracking the master branch. The legacy
    Kismet plugins should continue to work, but we will wait to track git
    master until development has stabalised.
  
 -- Dominic Spill <dominicgs@gmail.com>  01 Mar 2017 10:39:00 -0700   
 
 ubertooth (2015-10-R1) unstable; urgency=low
  * This is a minor bugfix and update release
  * Update ubertooth-scan to provide more consistent output
  * It is now possible to build receive only firmware
  * Host code versioning is more reliable
  * Python tools now support Python 3. Thanks Hannes!

  
 -- Dominic Spill <dominicgs@gmail.com>  12 Oct 2015 03:10:00 -0700   
 
 ubertooth (2015-09-R2) unstable; urgency=low
  * Basic Rate Bluetooth Improvements
      * Stability has improved in performance constrained environments
      * The "uh oh, buffer not emptied" issue no longer causes the tools to loop
        indefinitely
      * AFH support has greatly improved, UAP and CLK values are now found more
        quickly when AFH is used
  * Basic Rate PCAP support
      * ubertooth-rx and ubertooth-follow support a new PCAP file format
      * The new PCAP format is also supported with a new Wireshark plugin: btbredr
      * The PCAP format is still in the early stages, but we hope to have the plugin
        included in Wireshark later this year
      * The older btbb format is still produced by the Kismet plugins, they will
        be migrated over to the new format in the next release
  * Bluetooth Smart (BLE) Improvements
      * Improved packet reception
      * Promiscuous sniffing mode is more reliable
      * Promiscuous sniffing reports the recovered parameters to the host
  * Skateboard Hacking
      * Support for interfering with connections of BLE enabled skateboards
      * ubertooth-ego targets Yuneec E-go (not Bluetooth)
      * Always hack safely and responsibly!

  
 -- Dominic Spill <dominicgs@gmail.com>  05 Sep 2015 05:30:00 -0700   

ubertooth (2014-02-R2) unstable; urgency=low
  * Bluetooth Smart (Low Energy) Support
      * Promiscuous and follow modes
      * Pcap format packet logging
      * Pairing / encryption support when paired with crackle
      * Credit for BLE features goes to Mike Ryan
  * Unified host tool for monitoring Basic Rate
      * ubertooth-rx replaces -lap, -uap, -hop tools
      * Once UAP is discovered, ubertooth-rx automatically tries to find clock
        values and begin hopping
      * Thanks to Will Code for working on this
  * Survey tool - ubertooth-scan
      * Combining both Ubertooth and a standard Bluetooth dongle
      * Ubertooth scans for non-discoverable master devices
      * Dongle probes devices for piconet information and features
  * Cmake now used for the build system
      * Improves support for non-Linux operating systems
      * More sensible handling of dependencies
      * New build instructions
  * Packaging (Experimental)
      * Early stage support for packaging systems
      * libbtbb in Homebrew repository, Ubertooth coming soon
      * MacPorts availability is under test
      * Release already available in Pentoo
  * GitHub migration
      * libbtbb, Ubertooth and gr-bluetooth all hosted on GitHub
      * Allows for more open development and collaboration model
      * Already seeing an increase in issue reporting and pull requests


 -- Dominic Spill <dominicgs@gmail.com>  20 Feb 2014 06:06:00 -0700   


host/ubertooth-tools/src/ubertooth-rx.c
/*
 * Copyright 2010, 2011 Michael Ossmann
 *
 * This file is part of Project Ubertooth.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */

#include "ubertooth.h"
#include "ubertooth_callback.h"
#include <err.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>

static void usage()
{
	printf("ubertooth-rx - passive Classic Bluetooth discovery/decode\n");
	printf("\n");
	printf("Example usage:\n");
	printf("\tubertooth-rx -- sniff for all LAPs\n");
	printf("\tubertooth-rx -l <lap> -- calculate UAP for a given LAP\n");
	printf("\tubertooth-rx -l <lap> -u <uap> -- calculate clock and follow piconet\n");
	printf("\tubertooth-rx -z -t 20 -- survey mode: discover all LAPs+UAPs for 20 seconds\n");
	printf("\n");
	printf("Major modes:\n");
	printf("\t-l <LAP> to decode (6 hex) - if not specified sniff all LAPs\n");
	printf("\t-u <UAP> to decode (2 hex) - if not specified calculate UAP (requires LAP)\n");
	printf("\t-z Survey mode - discover and list piconets (implies -s, interrupt with ctrl-C)\n");
	printf("\t-i <filename> input file - if not specified use Ubertooth for live capture\n");
	printf("\n");
	printf("Configuration:\n");
	printf("\t-c <BT Channel> set a fixed bluetooth channel [Default: 39]\n");
	printf("\t-e max_ac_errors (default: %d, range: 0-4)\n", max_ac_errors);
	printf("\t-t <SECONDS> sniff timeout - 0 means no timeout [Default: 0]\n");
	printf("\n");
	printf("Output options:\n");
	printf("\t-r<filename> capture packets to PcapNG file\n");
	printf("\t-q<filename> capture packets to PCAP file\n");
	printf("\t-d<filename> dump packets to binary file\n");
	printf("\n");
	printf("Miscellaneous:\n");
	printf("\t-V print version information\n");
	printf("\t-U <0-7> set ubertooth device to use\n");
}

int main(int argc, char* argv[])
{
	int opt, have_lap = 0, have_uap = 0;
	int survey_mode = 0;
	int r;
	int timeout = 0;
	char* end;
	int ubertooth_device = -1;
	btbb_piconet* pn = NULL;
	uint32_t lap = 0;
	uint8_t uap = 0;
	uint16_t channel = 9999;

	ubertooth_t* ut = ubertooth_init();

	while ((opt=getopt(argc,argv,"hVi:l:u:U:d:e:r:sq:t:zc:")) != EOF) {
		switch(opt) {
		case 'i':
			infile = fopen(optarg, "r");
			if (infile == NULL) {
				printf("Could not open file %s\n", optarg);
				usage();
				return 1;
			}
			break;
		case 'l':
			lap = strtol(optarg, &end, 16);
			have_lap++;
			break;
		case 'u':
			uap = strtol(optarg, &end, 16);
			have_uap++;
			break;
		case 'U':
			ubertooth_device = atoi(optarg);
			break;
		case 'r':
			if (!ut->h_pcapng_bredr) {
				if (btbb_pcapng_create_file( optarg, "Ubertooth", &ut->h_pcapng_bredr )) {
					err(1, "create_bredr_capture_file: ");
				}
			}
			else {
				printf("Ignoring extra capture file: %s\n", optarg);
			}
			break;
		case 'q':
			if (!ut->h_pcap_bredr) {
				if (btbb_pcap_create_file(optarg, &ut->h_pcap_bredr)) {
					err(1, "btbb_pcap_create_file: ");
				}
			}
			else {
				printf("Ignoring extra capture file: %s\n", optarg);
			}
			break;
		case 'd':
			dumpfile = fopen(optarg, "w");
			if (dumpfile == NULL) {
				perror(optarg);
				return 1;
			}
			break;
		case 'e':
			max_ac_errors = atoi(optarg);
			break;
		case 's':
			fprintf(stderr, "sweep mode is now the default and the -s argument is deprecated\n");
			break;
		case 't':
			timeout = atoi(optarg);
			break;
		case 'z':
			++survey_mode;
			break;
		case 'c':
			channel = atoi(optarg);
			channel = channel + 2402;
			break;
		case 'V':
			print_version();
			return 0;
		case 'h':
		default:
			usage();
			return 1;
		}
	}

	if(survey_mode && (have_lap || have_uap)) {
		fprintf(stderr, "No address should be specified for survey mode\n");
		return 1;
	}

	if (infile == NULL) {
		r = ubertooth_connect(ut, ubertooth_device);
		if (r < 0) {
			usage();
			return 1;
		}

		r = ubertooth_check_api(ut);
		if (r < 0)
			return 1;
	}

	r = btbb_init(max_ac_errors);
	if (r < 0)
		return r;

	if(survey_mode) {
		// auto-flush stdout so that wrapper scripts work
		setvbuf(stdout, NULL, _IONBF, 0);
		btbb_init_survey();
	} else {
		if (have_lap) {
			pn = btbb_piconet_new();
			btbb_init_piconet(pn, lap);
			if (have_uap) {
				btbb_piconet_set_uap(pn, uap);
				if (infile == NULL)
					cmd_set_bdaddr(ut->devh, btbb_piconet_get_bdaddr(pn));
			}
			if (ut->h_pcapng_bredr) {
				btbb_pcapng_record_bdaddr(ut->h_pcapng_bredr,
							  (((uint32_t)uap)<<24)|lap,
							  have_uap ? 0xff : 0x00, 0);
			}
		} else if (have_uap) {
			fprintf(stderr, "Error: UAP but no LAP specified\n");
			usage();
			return 1;
		}
	}

	if (infile == NULL) {
		cmd_set_channel(ut->devh, channel);

		/* Clean up on exit. */
		register_cleanup_handler(ut, 0);

		if (timeout)
			ubertooth_set_timeout(ut, timeout);

		// init USB transfer
		r = ubertooth_bulk_init(ut);
		if (r < 0)
			return r;

		r = ubertooth_bulk_thread_start();
		if (r < 0)
			return r;

		// tell ubertooth to send packets
		r = cmd_rx_syms(ut->devh);
		if (r < 0)
			return r;

		// receive and process each packet
		while(!ut->stop_ubertooth) {
			ubertooth_bulk_receive(ut, cb_rx, pn);
		}

		ubertooth_bulk_thread_stop();

		ubertooth_stop(ut);
	} else {
		stream_rx_file(ut, infile, cb_rx, pn);
		fclose(infile);
	}

	if(survey_mode) {
		printf("Survey Results\n");
		while((pn=btbb_next_survey_result()) != NULL) {
			lap = btbb_piconet_get_lap(pn);
			if (btbb_piconet_get_flag(pn, BTBB_UAP_VALID)) {
				uap = btbb_piconet_get_uap(pn);
				/* Printable version showing that the NAP is unknown */
				printf("??:??:%02X:%02X:%02X:%02X\n", uap,
						(lap >> 16) & 0xFF, (lap >> 8) & 0xFF, lap & 0xFF);
			} else
				printf("??:??:??:%02X:%02X:%02X\n", (lap >> 16) & 0xFF,
					   (lap >> 8) & 0xFF, lap & 0xFF);
			//btbb_print_afh_map(pn);
		}
	}
	if(dumpfile != NULL)
		fclose(dumpfile);

	return 0;
}
